src/number_theory/legendre_symbol/mul_character.lean

```diff
@@ -0,0 +1,507 @@
+                   simp only [h1, dif_pos, units.coe_eq_one, map_one, is_unit_one], },
+  map_mul' :=
+  begin
+    intros x y,
+    by_cases hx : is_unit x,
+    { simp only [hx, is_unit.mul_iff, true_and, dif_pos],
+      by_cases hy : is_unit y,
+      { simp only [hy, dif_pos],
+        have hm : (is_unit.mul_iff.mpr ⟨hx, hy⟩).unit = hx.unit * hy.unit := units.eq_iff.mp rfl,
+        rw [hm, map_mul],
+        norm_cast, },
+      { simp only [hy, not_false_iff, dif_neg, mul_zero], }, },
+    { simp only [hx, is_unit.mul_iff, false_and, not_false_iff, dif_neg, zero_mul], },
+  end ,
+  map_nonunit' := by { intros a ha, simp only [ha, not_false_iff, dif_neg], }, }
+
+lemma of_unit_hom_eval_def (f : Rˣ →* R'ˣ) (a : R) :
+  of_unit_hom f a = if ha : is_unit a then f ha.unit else 0 := rfl
+
+lemma of_unit_hom_eval (f : Rˣ →* R'ˣ) {a : R} (ha : is_unit a) :
+  of_unit_hom f a = f ha.unit :=
+by simp only [ha, of_unit_hom_eval_def, dif_pos]
+
+lemma of_unit_hom_eval' (f : Rˣ →* R'ˣ) (a : Rˣ) :
+  of_unit_hom f a = f a :=
+by simp only [of_unit_hom_eval_def, units.is_unit, is_unit.unit_of_coe_units, dite_eq_ite, if_true]
+
+/-- The equivalence between multiplicative characters and homomorphisms of unit groups. -/
+noncomputable
+def equiv_to_unit_hom : mul_char R R' ≃ (Rˣ →* R'ˣ) :=
+{ to_fun := to_unit_hom,
+  inv_fun := of_unit_hom,
+  left_inv :=
+  by { intro χ, ext x hx, rw [of_unit_hom_eval _ hx, to_unit_hom_eval, is_unit.unit_spec] },
+  right_inv :=
+  by { intro f, ext x, rw [to_unit_hom_eval', ← of_unit_hom_eval' f x, is_unit.unit_spec], } }
+
+/-!
+### Commutative group structure on multiplicative characters
+
+The multiplicative characters `R → R'` form a commutative group.
+-/
+
+@[protected]
+lemma map_one (χ : mul_char R R') : χ (1 : R) = 1 :=
+χ.map_one'
+
+/-- If the domain has a zero (and is nontrivial), then `χ 0 = 0`. -/
+@[protected]
+lemma map_zero {R : Type u} [comm_monoid_with_zero R] [nontrivial R] (χ : mul_char R R') :
+  χ (0 : R) = 0 :=
+by rw [map_nonunit χ not_is_unit_zero]
+
+noncomputable
+instance has_one : has_one (mul_char R R') := ⟨trivial R R'⟩
+
+noncomputable
+instance inhabited : inhabited (mul_char R R') := ⟨1⟩
+
+/-- Evaluation of the trivial character -/
+lemma one_eval {a : R} : (1 : mul_char R R') a = if (is_unit a) then 1 else 0 :=
+rfl
+
+@[simp]
+lemma one_eval_of_is_unit {a : R} (ha : is_unit a) :
+  (1 : mul_char R R') a = 1 :=
+by simp only [one_eval, ha, if_true]
+
+@[simp]
+lemma one_eval_of_not_is_unit {a : R} (ha : ¬ is_unit a) :
+  (1 : mul_char R R') a = 0 :=
+by simp only [one_eval, ha, if_false]
+
+/-- Multiplication of multiplicative characters. (This needs the target to be commutative.) -/
+def mul (χ χ' : mul_char R R') : mul_char R R' :=
+{ map_nonunit' :=
+  begin
+    intros a ha,
+    simp only,
+    have h : (χ.to_monoid_hom : R → R') = χ := rfl,
+    rw [monoid_hom.to_fun_eq_coe, monoid_hom.mul_apply, h, map_nonunit χ ha, zero_mul],
+  end,
+  ..χ.to_monoid_hom * χ'.to_monoid_hom }
```

Comment:
By manually supplying a value for the data-carrying fields (here `to_fun`), we can control the unfolding behaviour and thereby simplify the proofs:
```suggestion
{ to_fun := χ * χ',
  map_nonunit' := λ a ha, by simp [map_nonunit χ ha],
  ..χ.to_monoid_hom * χ'.to_monoid_hom }
```